contracts/vortex/CarbonVortex.sol

```diff
@@ -0,0 +1,262 @@
+        uint256 len = tokens.length;
+
+        // allocate balances array for the tokens
+        uint256[] memory balances = new uint256[](len);
+        // allocate array for the reward amounts for caller
+        uint256[] memory rewardAmounts = new uint256[](len);
+
+        // withdraw fees, load balances and reward amounts
+        for (uint256 i = 0; i < len; i = uncheckedInc(i)) {
+            // withdraw token fees
+            _carbonController.withdrawFees(tokens[i], type(uint256).max, address(this));
+            // get token balance
+            balances[i] = tokens[i].balanceOf(address(this));
+            // get reward amount for token
+            rewardAmounts[i] = MathEx.mulDivF(balances[i], _rewardsPercentagePPM, PPM_RESOLUTION);
+        }
+
+        // convert tokens to BNT
+        for (uint256 i = 0; i < len; i = uncheckedInc(i)) {
+            Token token = tokens[i];
+            // skip token if no token balance
+            if (balances[i] == 0) {
+                continue;
+            }
+            // check if token is BNT - we don't need to swap in this case
+            if (token == _bnt) {
+                continue;
+            }
+
+            // get trade amount for token
+            uint256 tradeAmount = balances[i] - rewardAmounts[i];
+
+            // approve tokens for trading on Bancor Network V3
+            _setAllowance(token, tradeAmount);
+
+            uint256 val = token.isNative() ? tradeAmount : 0;
+
+            // swap token to BNT using Bancor Network V3
+            _bancorNetwork.tradeBySourceAmount{ value: val }(token, _bnt, tradeAmount, 1, block.timestamp, address(0));
+        }
+
+        // allocate rewards to caller and burn the rest
+        _allocateRewards(msg.sender, tokens, rewardAmounts);
+    }
+
+    /**
+     * @dev allocates the rewards to caller and burns the rest
+     */
+    function _allocateRewards(address sender, Token[] calldata tokens, uint256[] memory rewardAmounts) private {
+        // transfer the rewards to caller
+        uint256 len = tokens.length;
+        for (uint256 i = 0; i < len; i = uncheckedInc(i)) {
+            Token token = tokens[i];
+            uint256 rewardAmount = rewardAmounts[i];
+            if (rewardAmount == 0) {
+                continue;
+            }
+            if (token.isNative()) {
+                // using a regular transfer here could revert due to exceeding the 2300 gas limit
+                // which is why we're using call instead (via sendValue)
+                payable(sender).sendValue(rewardAmount);
+            } else {
+                token.safeTransfer(sender, rewardAmount);
+            }
+        }
+
+        // get the burn amount
+        uint256 burnAmount = _bnt.balanceOf(address(this));
+
+        // add to the total burnt amount
+        if (burnAmount > 0) {
+            _totalBurnt += burnAmount;
+        }
+
+        // burn the tokens
+        _bnt.safeTransfer(Token.unwrap(_bnt), burnAmount);
```

Comment:
This can be inside the `if (burnAmount > 0)` I believe

Reply:
Well, you save both the function call (to `safeTransfer`), and the additional check (of `burnAmount > 0`) inside that function.

Reply:
Saves about 50 gas, but it's still something. Changed.